readthedocs/projects/notifications.py

```diff
@@ -1,28 +1,116 @@
-"""Project notifications."""
+"""Notifications related to projects."""
 
-from django.urls import reverse
-from django.utils.translation import gettext_lazy as _
-from messages_extends.constants import ERROR_PERSISTENT
+from django.utils.translation import gettext_noop as _
 
-from readthedocs.notifications import Notification, SiteNotification
-from readthedocs.notifications.constants import REQUIREMENT
+from readthedocs.notifications.constants import ERROR, INFO
+from readthedocs.notifications.messages import Message, registry
+from readthedocs.projects.exceptions import (
+    ProjectConfigurationError,
+    RepositoryError,
+    UserFileNotFound,
+)
 
-
-class ResourceUsageNotification(Notification):
-    name = "resource_usage"
-    context_object_name = "project"
-    subject = "Builds for {{ project.name }} are using too many resources"
-    level = REQUIREMENT
-
-
-class EmailConfirmNotification(SiteNotification):
-    failure_level = ERROR_PERSISTENT
-    failure_message = _(
-        "Your primary email address is not verified. "
-        'Please <a href="{{account_email_url}}">verify it here</a>.',
-    )
-
-    def get_context_data(self):
-        context = super().get_context_data()
-        context.update({"account_email_url": reverse("account_email")})
-        return context
+MESSAGE_PROJECT_SKIP_BUILDS = "project:invalid:skip-builds"
+messages = [
+    Message(
+        id=MESSAGE_PROJECT_SKIP_BUILDS,
+        header=_("Build skipped for this project"),
+        body=_(
+            """
+            Your project is currently disabled for abuse of the system.
+            Please make sure it isn't using unreasonable amounts of resources or triggering lots of builds in a short amount of time.
+            Please <a href="mailto:{SUPPORT_EMAIL}">contact support</a> to get your project re-enabled.
+            """
+        ),
+        type=INFO,
+    ),
+    Message(
+        id=RepositoryError.GENERIC,
+        header=_("Error while cloning the repository"),
+        body=_(
+            """
+            There was a problem cloning your repository.
+            Please check the command output for more information.
+            """
+        ),
+        type=ERROR,
+    ),
+    Message(
+        id=RepositoryError.DUPLICATED_RESERVED_VERSIONS,
+        header=_("Duplicated reserved versions"),
+        body=_(
+            """
+            You can not have two versions with the name latest or stable.
+            Ensure you don't have both a branch and a tag with this name.
+            """
+        ),
+        type=ERROR,
+    ),
+    Message(
+        id=RepositoryError.FAILED_TO_CHECKOUT,
+        header=_("Error while checking out the repository"),
+        body=_(
+            """
+            Failed to checkout revision: <code>{revision}</code>
```

Comment:
Hmm, are we including these messages directly in the HTML? They could probably be abused to inject malicious content, since we are using `.format` to render the body, as opposed to something like Django's template engine that escapes HTML by default.

Reply:
Yes. Including the text directly in HTML was one of the initial requirements for this project.

Currently, we are not, and we _should not_ in the future, format data that comes from the user. All the data formatted comes from the backend and should be safe to use inside `.format` since we have full control of it.

Let me know if you found a particular case that you are worried about and we can take a deeper look at it.

Reply:
Well, in this case the name of the revision comes from the user, an easier one to inject is from the config file

https://github.com/readthedocs/readthedocs.org/blob/e7b83e0c9a5da57236650843bda8025d7470c189/readthedocs/config/notifications.py#L108


```yaml
version: 2

<strong>hello</strong>: world
```

We should escape for HTML all values, otherwise this is vulnerable to stored XSS.

Reply:
PR at https://github.com/readthedocs/readthedocs.org/pull/10996